### PR TITLE
Variadic when and wait functions

### PR DESCRIPTION
**Pull request by hkaiser**
This patch proposes to change the implementation of all `wait_xxx` and `when_xxx` free functions to use variadics. This introduces API breaking changes for some of those, the sequence of the parameter have been changed to ensure that all variadic arguments come last.

This patch also adds the missing documentation for `when_each` and `wait_each`.

All tests pass: http://hermione.cct.lsu.edu/builders/hpx_gcc_4_9_boost_1_56_debian_x86_64_debug/builds/217

### CODE REVIEW

File: hpx/lcos/wait_some.hpp [hunk 1]
```diff
@@ -468,69 +466,54 @@ namespace hpx { namespace lcos
...
-        typedef HPX_STD_TUPLE<
-            BOOST_PP_ENUM(N, HPX_WAIT_SOME_SHARED_STATE_FOR_FUTURE, _)>
-            result_type;
+        typedef hpx::util::tuple<
+                typename lcos::detail::shared_state_ptr_for<Ts>::type...
+            > result_type;
 
-        result_type lazy_values_(
-            BOOST_PP_ENUM(N, HPX_WAIT_SOME_GET_SHARED_STATE, _));
+        result_type lazy_values_(lcos::detail::get_shared_state(ts)...);
```
Comment (K-ballo) [inline hpx/lcos/wait_some.hpp]: Doesn't this cause a pattern-matching ICE?

File: hpx/util/functional/boolean_ops.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ...Ts>
+    struct any_of
+      : boost::mpl::fold<
+            util::tuple<Ts...>, boost::mpl::false_,
+            boost::mpl::or_<boost::mpl::_1, boost::mpl::_2>
+        >::type
+    {};
+}}}
```
Comment (K-ballo) [inline hpx/util/functional/boolean_ops.hpp]: These implementations are quite inefficient at compile-time, but can be replaced later.

**Comment by K-ballo** [comment]
LGTM

File: hpx/lcos/wait_some.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/lcos/wait_some.hpp]: Interesting - it should. Apparently this is not tested anywhere.

File: hpx/util/functional/boolean_ops.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/util/functional/boolean_ops.hpp]: I tried to implement it directly, but failed :(   Any suggestions are welcome.

**Pull request merged**